src/writers/writer.cpp

```diff
@@ -8,255 +8,451 @@
+            paths.pop();
+            for (auto j = 0; j < n_shards; ++j) {
+                paths.push(path / std::to_string(j));
+            }
+        }
+    }
 
-    return !failure;
+    return make_files_(paths, files);
 }
 
 bool
-zarr::FileCreator::create_c_dirs_(int n_c)
+zarr::FileCreator::make_dirs_(std::queue<fs::path>& dir_paths)
 {
-    std::latch latch(n_c);
-    std::atomic<bool> failure{ false };
-    for (auto c = 0; c < n_c; ++c) {
+    if (dir_paths.empty()) {
+        return true;
+    }
+
+    std::atomic<bool> success = true;
+
+    const auto n_dirs = dir_paths.size();
+    std::latch latch(n_dirs);
+
+    for (auto i = 0; i < n_dirs; ++i) {
+        const auto dirname = dir_paths.front();
+        dir_paths.pop();
+
         thread_pool_->push_to_job_queue(
-          std::move([base = base_dir_, c, &latch, &failure](std::string& err) {
+          [dirname, &latch, &success](std::string& err) -> bool {
               try {
-                  const auto path = base / std::to_string(c);
-                  if (fs::exists(path)) {
-                      EXPECT(fs::is_directory(path),
-                             "%s must be a directory.",
-                             path.c_str());
-                  } else if (!failure) {
-                      EXPECT(fs::create_directories(path),
-                             "Failed to create directory: %s",
-                             path.c_str());
+                  if (fs::exists(dirname)) {
+                      EXPECT(fs::is_directory(dirname),
+                             "'%s' exists but is not a directory.",
+                             dirname.c_str());
+                  } else if (success) {
+                      EXPECT(fs::create_directories(dirname),
+                             "Not creating directory '%s': another job failed.",
+                             dirname.c_str());
                   }
               } catch (const std::exception& exc) {
                   char buf[128];
                   snprintf(buf,
                            sizeof(buf),
-                           "Failed to create directory: %s",
+                           "Failed to create directory '%s': %s.",
+                           dirname.string().c_str(),
+                           exc.what());
+                  err = buf;
+                  success = false;
+              } catch (...) {
+                  char buf[128];
+                  snprintf(buf,
+                           sizeof(buf),
+                           "Failed to create directory '%s': (unknown).",
+                           dirname.string().c_str());
+                  err = buf;
+                  success = false;
+              }
+
+              latch.count_down();
+              return success;
+          });
+
+        dir_paths.push(dirname);
+    }
+
+    latch.wait();
+
+    return success;
+}
+
+bool
+zarr::FileCreator::make_files_(std::queue<fs::path>& file_paths,
```

Comment:
Unlike `make_dirs_` this mutates `file_paths` by consuming the queue (i.e. we only `pop` here with no corresponding `push`). I think it's worth mentioning that distinction in the header docs for these methods.

Reply:
Can do